Parallelize iSCSI tests

jsafrane opened the pull request:
iSCSI target (=the server) is implemented in Linux kernel. The "iSCSI server" pod is not a real server, it just configures the kernel on the host. In order to run iSCSI tests in parallel, we need to be able to
run multiple such pods on a single node, serving different LUNs to different tests.

The "server pod" must run with HostNetwork=true to achieve that. Each pod then creates its own IQN with namespace name, so it can't collide with other server pods running in another namespaces on the same node.

Fixes: #76217

**What type of PR is this?**
/kind cleanup
/sig storage

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

cc @rootfs @gnufied

jsafrane commented:
Bumped iSCSI image version to 2.0 because it's incompatible with previous 1.0 version.

jsafrane committed:
hack

in VM:
make all WHAT=volume/iscsi
(+ break after amd64)

jsafrane commented:
@ixdy, you build my last changes in test images, I am not sure what's the current process. Can you please re-build and push the image? Or at least approve this PR?

No current test uses the image, so it does not really matter if it's merged before or after the containers are rebuilt. I'll try to add a new prow job at least for iSCSI so it's tested.

gnufied commented on test/images/volume/iscsi/VERSION:
```diff
@@ -1 +1 @@
-1.0
+2.0
```
Whats there in `block.tar.gz` file jan?

jsafrane replied:
See README.md:

> block.tar.gz is a small ext2 filesystem created by `create_block.sh` (run as root!)

I had to update it from "Hello from iSCSI" to "Hello from iscsi" for a mount test to pass, as it expects test driver name there.

gnufied commented on test/e2e/framework/volume_util.go:
```diff
@@ -174,21 +179,31 @@ func NewGlusterfsServer(cs clientset.Interface, namespace string) (config Volume
+			// iSCSI container needs to configure kernel
+			"/sys/kernel": "/sys/kernel",
+			// iSCSI source "block devices" must be available on the host
+			"/srv/iscsi": "/srv/iscsi",
```
Do we need bi-directional mount propagation here?

jsafrane replied:
I think it's not necessary. `targetcli` does not mount anything by itself.

The test "server" pod just puts block files there (block.tar.gz) and iSCSI target implementation in kernel finds them on the same path on the host.

ixdy commented:
/assign @listx

jsafrane commented:
/retest

wongma7 commented on test/e2e/framework/volume_util.go:
```diff
@@ -174,21 +179,31 @@ func NewGlusterfsServer(cs clientset.Interface, namespace string) (config Volume
 			// iSCSI container needs to insert modules from the host
 			"/lib/modules": "/lib/modules",
+			// iSCSI container needs to configure kernel
+			"/sys/kernel": "/sys/kernel",
```
what happens if this is not mounted? (just curious why we did not need it before)

jsafrane replied:
iscsiadm was not happy and threw random errors.

wongma7 commented:
lgtm. targetcli commands in run_iscsi_target.sh look correct

childsb commented:
/lgtm

jsafrane commented:
@listx, can you please approve and build/push the image to gcr.io?

listx commented:
/lgtm
/approve

jsafrane commented:
/assign @saad-ali 
for approval

listx commented:
/approve

jsafrane commented:
/assign @oomichi @andrewsykim 

for approval of `test/e2e/framework`

andrewsykim commented on test/e2e/framework/volume_util.go:
```diff
@@ -174,21 +179,31 @@ func NewGlusterfsServer(cs clientset.Interface, namespace string) (config Volume
+			// iSCSI source "block devices" must be available on the host
+			"/srv/iscsi": "/srv/iscsi",
 		},
-		ServerReadyMessage: "Configuration restored from /etc/target/saveconfig.json",
```
Is removing the ServerReadyMessage check intentional here?

jsafrane replied:
Good catch, replaced with `"iscsi target started"`.

oomichi commented on test/e2e/framework/volume_util.go:
```diff
@@ -174,21 +179,31 @@ func NewGlusterfsServer(cs clientset.Interface, namespace string) (config Volume
 
-// iSCSI-specific wrapper for CreateStorageServer.
-func NewISCSIServer(cs clientset.Interface, namespace string) (config VolumeTestConfig, pod *v1.Pod, ip string) {
+// iSCSI-specific wrapper for CreateStorageServer. Note that there can be only
```
Now we are fixing golint failures under `test/e2e/framework`.
So please add "NewISCSIServer is an" for avoiding golint failures.

jsafrane replied:
Fixed

oomichi commented on test/e2e/framework/volume_util.go:
```diff
@@ -174,21 +179,31 @@ func NewGlusterfsServer(cs clientset.Interface, namespace string) (config Volume
+// one SCSI server running on a node, [Serial] is highly advised.
+func NewISCSIServer(cs clientset.Interface, namespace string) (config VolumeTestConfig, pod *v1.Pod, ip, iqn string) {
+	// Generate cluster-wide unique IQN
+	iqn = fmt.Sprintf(iSCSIIQNTemplate, namespace)
```
+1 for replacing static `"iqn.2003-01.org.linux-iscsi.f21.x8664:sn.4b0aae584f7c"`

oomichi reviewed (commented):
Thanks for doing this, just few comments

jsafrane committed:
Parallelize iSCSI tests

iSCSI target (=the server) is implemented in Linux kernel. The "iSCSI
server" pod is not a real server, it just configures the kernel on the
host. In order to run iSCSI tests in parallel, we need to be able to
run multiple such pods on a single node, serving different LUNs to
different tests.

The "server pod" must run with HostNetwork=true to achieve that.
Each pod then creates its own IQN with namespace name, so it can't
collide with other server pods running in another namespaces on the same
node.

jsafrane commented:
@oomichi @andrewsykim I fixed all your comments, PTAL again

oomichi commented:
@jsafrane 
Thanks for updating, looks good for me now.
/lgtm

andrewsykim commented:
/approve
